pkg/client/query/cache/memory.go

```diff
@@ -0,0 +1,362 @@
+		return zero, ErrCacheMiss.Wrapf("key: %s", key)
+	}
+
+	isTTLEnabled := c.config.ttl > 0
+	isCacheItemExpired := time.Since(cachedItem.cachedAt) > c.config.ttl
+	if isTTLEnabled && isCacheItemExpired {
+		// DEV_NOTE: Intentionally not pruning here to improve concurrent speed;
+		// otherwise, the read lock would be insufficient. The value will be
+		// overwritten by the next call to Set(). If usage is such that values
+		// aren't being subsequently set, maxKeys (if configured) will eventually
+		// cause the pruning of values with expired TTLs.
+		return zero, ErrCacheMiss.Wrapf("key: %s", key)
+	}
+
+	return cachedItem.value, nil
+}
+
+// GetAtHeight retrieves the value from the cache with the given key, at the given
+// height. If a value is not found for that height, the value at the nearest previous
+// height is returned. If the cache is not configured for historical mode, it returns
+// an error.
+func (c *inMemoryCache[T]) GetAtHeight(key string, getHeight int64) (T, error) {
+	var zero T
+
+	if !c.config.historical {
+		return zero, ErrHistoricalModeNotEnabled
+	}
+
+	c.valuesMu.RLock()
+	defer c.valuesMu.RUnlock()
+
+	valueHistory := c.valueHistories[key]
+	var nearestCachedHeight int64 = -1
+	for _, cachedHeight := range valueHistory.sortedDescHeights {
+		if cachedHeight <= getHeight {
+			nearestCachedHeight = cachedHeight
+			// DEV_NOTE: Since the list is sorted in descending order, once we
+			// encounter a cachedHeight that is less than or equal to getHeight,
+			// all subsequent cachedHeights SHOULD also be less than or equal to
+			// getHeight.
+			break
+		}
+	}
+
+	if nearestCachedHeight == -1 {
+		return zero, ErrCacheMiss.Wrapf("key: %s, height: %d", key, getHeight)
+	}
+
+	value, exists := valueHistory.heightMap[nearestCachedHeight]
+	if !exists {
+		return zero, ErrCacheInternal.Wrapf("failed to load historical value for key: %s, height: %d", key, getHeight)
+	}
+
+	if c.config.ttl > 0 && time.Since(value.cachedAt) > c.config.ttl {
+		// DEV_NOTE: Intentionally not pruning here to improve concurrent speed;
+		// otherwise, the read lock would be insufficient. The value will be pruned
+		// in the subsequent call to SetAtHeight() after c.config.numHistoricalValues
+		// blocks have elapsed. If usage is such that historical values aren't being
+		// subsequently set, numHistoricalBlocks (if configured) will eventually
+		// cause the pruning of historical values with expired TTLs.
+		return zero, ErrCacheMiss.Wrapf("key: %s, height: %d", key, getHeight)
+	}
+
+	return value.value, nil
+}
+
+// Set adds or updates the value in the cache for the given key. If the cache is
+// configured for historical mode, it will store the value at the latest **known**
+// height, which is only updated on calls to SetAtHeight, and therefore is not
+// guaranteed to be the current height.
+func (c *inMemoryCache[T]) Set(key string, value T) error {
+	if c.config.historical {
+		return c.SetAtHeight(key, value, c.latestHeight.Load())
```

Comment:
I think this is more about the specific behavior of `Set()` in historical mode. This is indeed an edge case, and I don't think we should find ourselves doing this unless for a very specific reason. In such a case, I think it's more accurate to say that instead of `fetchValue()` updating the latest version (height), it would be an intermediate call to `SetAsOfVersion()` where the version is newer than the current latest version. I think this is the scenario I outlined above: https://github.com/pokt-network/poktroll/pull/994#discussion_r1889842645.

Reply:
As discussed, I've disabled `#Set()` in historical mode.